modules/mcagar/src/main/ts/ephox/mcagar/api/TinyHooks.ts

```diff
@@ -0,0 +1,124 @@
+import { after, afterEach, before } from '@ephox/bedrock-client';
+import { Arr, Fun, Optional, Strings, Type } from '@ephox/katamari';
+import { Insert, Remove, SugarBody, SugarElement, SugarShadowDom } from '@ephox/sugar';
+import 'tinymce';
+import { Editor as EditorType } from '../alien/EditorTypes';
+import * as Loader from '../loader/Loader';
+import { setTinymceBaseUrl } from '../loader/Urls';
+
+export interface Hook<T extends EditorType> {
+  readonly editor: () => T;
+}
+
+export interface ShadowRootHook<T extends EditorType> extends Hook<T> {
+  readonly shadowRoot: () => SugarElement<ShadowRoot>;
+}
+
+const setupBaseUrl = (tinymce: any, settings: Record<string, any>): void => {
+  if (settings.base_url) {
+    setTinymceBaseUrl(tinymce, settings.base_url);
+  } else if (!Type.isString(tinymce.baseURL) || !Strings.contains(tinymce.baseURL, '/project/')) {
+    setTinymceBaseUrl(tinymce, '/project/node_modules/tinymce');
+  }
+};
+
+const setupHooks = <T extends EditorType = EditorType>(
+  settings: Record<string, any>,
+  setupModules: Array<() => void>,
+  focusOnInit: boolean,
+  createElement: () => Optional<SugarElement>,
+  teardown = Fun.noop
+): Hook<T> => {
+  let editor: T;
+  let teardownEditor: () => void = Fun.noop;
+  let hasFailure = false;
+
+  before((done) => {
+    Arr.each(setupModules, Fun.call);
+    Loader.setup({
+      preInit: setupBaseUrl,
+      run: (ed, success) => {
+        editor = ed;
+        teardownEditor = success;
+        if (focusOnInit) {
+          editor.focus();
+        }
+        done();
+      },
+      success: Fun.noop,
+      failure: done
+    }, settings, createElement());
+  });
+
+  afterEach(function () {
```

Comment:
Arrow function?

Reply:
Again, can't be as it uses `this` (it's a Mocha equivalence thing).